File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) 
{
    map<string, int> wMap;
    int totalDay = 0;
    int cnt = 0;
    for(int n : number)
    {
        totalDay += n;
    }
    
    for(int i = 0; i<want.size(); i++)
    {
        wMap[want[i]] = number[i];
    }
    
    int k = 0;
    while(k + totalDay <= discount.size())
    {
        map<string, int> tmp = wMap;
        bool isBool = false;
        for(int i=k; i<k + totalDay; i++)
        {
            tmp[discount[i]]--;
        }
        
        
        for(string w : want)
        {
            if(tmp[w] != 0)
            {
                isBool = true;
                break;
            }
        }
        if(!isBool) { cnt++; }
        k++;
    }
    return cnt;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.cs
using System;
using System.Collections.Generic;

public class Solution 
{
    public int solution(string[] want, int[] number, string[] discount) 
    {
        // 사전에 등록
        Dictionary<string, int> wantDic = new Dictionary<string, int>();
        for(int i=0; i<want.Length; i++)
        {
            wantDic[want[i]] = number[i];
        }
        
        int count = 0;
        for(int i=0; i<=discount.Length - 10; i++)
        {
            Dictionary<string, int> copy = new Dictionary<string, int>(wantDic);
            for(int j=i; j<i+10; j++)
            {
                if(!copy.ContainsKey(discount[j]))
                {
                    continue;
                }
                copy[discount[j]]--;
            }
            
            bool b = true;
            foreach(var v in copy)
            {
                if(v.Value != 0)
                {
                    b = false;
                    break;
                }
            }
            
            if(b)
            {
                count++;
            }
        }
        return count;
    }
}